keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,262 @@
+authors:
+  - "@freehan"
+owning-sig: sig-network
+reviewers:
+  - "@bowei"
+  - "@thockin"
+  - "@wojtek-t"
+  - "@johnbelamaric"
+approvers:
+  - "@bowei"
+  - "@thockin"
+creation-date: 2019-06-01
+last-updated: 2019-06-01
+status: implementable
+see-also:
+  - "https://docs.google.com/document/d/1sLJfolOeEVzK5oOviRmtHOHmke8qtteljQPaDUEukxY/edit#"
+---
+# EndpointSlice API 
+
+## Summary 
+
+This KEP was converted from the [original proposal doc][original-doc]. The current  [Core/V1 Endpoints API][v1-endpoints-api] comes with severe performance/scalability drawbacks affecting multiple components in the control-plane (apiserver, etcd, endpoints-controller, kube-proxy). 
+This doc proposes a new EndpointSlice API aiming to replace Core/V1 Endpoints API for most internal consumers, including kube-proxy.
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+## Motivation
+
+In the current Endpoints API, one object instance contains all the individual endpoints of a service. Whenever a single pod in a service is added/updated/deleted, the whole Endpoints object (even when the other endpoints didn't change) is re-computed, written to storage (etcd) and sent to all watchers (e.g. kube-proxy). This leads to 2 major problems:
+
+- Storing multiple megabytes of endpoints puts strain on multiple parts of the system due to not having a paging system and a monolithic watch/storage design. [The max number of endpoints is bounded by the K8s storage layer (etcd)][max-object-size]], which has a hard limit on the size of a single object (1.5MB by default). That means attempts to write an object larger than the limit will be rejected. Additionally, there is a similar limitation in the watch path in Kubernetes apiserver. For a K8s service, if its Endpoints object is too large, endpoint updates will not be propagated to kube-proxy(s), and thus iptables/ipvs won’t be reprogrammed.
+- [Performance degradation in large k8s deployments.][perf-degrade] Not being able to efficiently read/update individual endpoint changes can lead to (e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If one consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+### Goal
+
+- Support tens of thousands of backend endpoints in a single service on cluster with thousands of nodes.
+- Leave room for foreseeable extension:
+  - Support multiple IPs per pod
+  - More endpoint states than Ready/NotReady
+  - Dynamic endpoint subsetting
+  
+### Non-Goal
+- Change functionality provided by K8s V1 Service API.
+- Provide better load balancing for K8s service backends.
+
+## Proposal
+
+### EndpointSlice API
```

Comment:
Slice implies a piece of a whole, rather than a whole (as in Set).  It's not horrible, but I wanted an anchor point for any debate :)

Reply:
I'm curious how this specific set is selected? I'm not sure I understand what this means or how this works.

Reply:
It's just a chunk.  A semi-random subset of a whole.

Reply:
You mentioned a crossed-correlation between `EndPoints` and `Ports`. How are these separated in slices? I think the API is sort of broken if objects can't live independently. If we do accept that for trade-offs reasons, then it needs to be very explicit. I like chunk more than slice because it conveys the fact that a chunk doesn't make sense on its own (all chunks are needed in order to use it).

Reply:
For a given slice the whole Ports[*] block applies to all Endpoints[*].  So pathologically, if every endpoint has a different port number, they would all get slices of length 1.